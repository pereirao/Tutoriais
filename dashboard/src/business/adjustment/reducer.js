//@flow

import { combineReducers } from 'redux';
import type { Adjustment } from './index';
import type { Action } from '../store';

type ById = {[id: string]: Adjustment};

export const byIdReducer = (state: ById = {}, action: Action) => {
  switch (action.type){
    case 'ADJUSTMENT:FETCH__SUCCESS':
      return { ...state, ...action.payload.entities.adjustment };
    default:
      return state;
  }
};

export const isFetchingReducer = (state: boolean = false, action: Action) => {
  switch (action.type){
    case 'ADJUSTMENT:FETCH__LOADING':
      return true;
    case 'ADJUSTMENT:FETCH__SUCCESS':
    case 'ADJUSTMENT:FETCH__ERROR':
      return false;
    default:
      return state;
  }
};

export type LocalState = {
  by_id: ById,
  is_fetching: boolean
};

const adjustmentReducer = combineReducers({
  by_id: byIdReducer,
  is_fetching: isFetchingReducer
});

export default adjustmentReducer;
