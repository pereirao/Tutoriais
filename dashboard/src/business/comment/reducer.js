//@flow

import { combineReducers } from 'redux';
import type { Comment } from './index';
import type { Action } from '../store';

type ById = {[id: string]: Comment};

export const byIdReducer = (state: ById = {}, action: Action) => {
  switch (action.type){
    case 'COMMENT:FETCH__SUCCESS':
    case 'COMMENT:ADD__SUCCESS':
      return { ...state, ...action.payload.entities.comment };
    default:
      return state;
  }
};

export const isFetchingReducer = (state: boolean = false, action: Action) => {
  switch (action.type){
    case 'COMMENT:FETCH__LOADING':
    case 'COMMENT:ADD__LOADING':
      return true;
    case 'COMMENT:FETCH__SUCCESS':
    case 'COMMENT:ADD__SUCCESS':
    case 'COMMENT:FETCH__ERROR':
    case 'COMMENT:ADD__ERROR':
      return false;
    default:
      return state;
  }
};

export type LocalState = {
  by_id: ById,
  is_fetching: boolean
};

const commentReducer = combineReducers({
  by_id: byIdReducer,
  is_fetching: isFetchingReducer
});

export default commentReducer;
