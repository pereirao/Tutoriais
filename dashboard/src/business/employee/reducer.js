// @flow

import { combineReducers } from 'redux';
import _ from 'lodash';
// import I18n from '../../localization';
import type { Order } from './index';
import type { Action } from '../store';

type ById = {[id: number]: Order};

export const byIdReducer = (state: ById = {}, action: Action) => {
  switch (action.type){
    case 'EMPLOYEE:FETCH__SUCCESS':
      return action.payload.entities.employee || {};
    case 'EMPLOYEE:CREATE__SUCCESS':
    case 'EMPLOYEE:UPDATE__SUCCESS':
      return {...state, ...action.payload.entities.employee};
    case 'EMPLOYEE:DESTROY__SUCCESS':
      return _.omit(state, action.payload.result);
    default:
      return state;
  }
};

export const allIdsReducer = (state: Array<string> = [], action: Action) => {
  switch (action.type){
    case 'EMPLOYEE:FETCH__SUCCESS':
      return _.uniq([...state, ...action.payload.result]);
    case 'EMPLOYEE:CREATE__SUCCESS':
      return [...state, action.payload.result];
    case 'EMPLOYEE:DESTROY__SUCCESS':
      return action.payload.result;
    default:
      return state;
  }
};

export const isFetchingReducer = (state: boolean = false, action: Action) => {
  switch (action.type){
    case 'EMPLOYEE:FETCH__LOADING':
      return true;
    case 'EMPLOYEE:FETCH__SUCCESS':
    case 'EMPLOYEE:FETCH__ERROR':
      return false;
    default:
      return state;
  }
};

export const updatingIdReducer = (state: number = null, action: Action) => {
  switch (action.type){
    case 'EMPLOYEE:UPDATE__LOADING':
      return action.meta.id;
    case 'EMPLOYEE:UPDATE__SUCCESS':
    case 'EMPLOYEE:UPDATE__ERROR':
      return null;
    default:
      return state;
  }
};

export const isCreatingReducer = (state: boolean = false, action: Action) => {
  switch (action.type){
    case 'EMPLOYEE:CREATE__LOADING':
      return true;
    case 'EMPLOYEE:CREATE__SUCCESS':
    case 'EMPLOYEE:CREATE__ERROR':
      return false;
    default:
      return state;
  }
};

export const successfullyCreatedReducer = (state: boolean = false, action: Action) => {
  switch (action.type){
    case 'EMPLOYEE:CREATE__SUCCESS':
      return true;
    case 'EMPLOYEE:RESET_NEW_EMPLOYEE_FORM':
    case 'EMPLOYEE:CREATE__ERROR':
      return false;
    default:
      return state;
  }
};

export const createEmployeeErrorsReducer = (state: Array<string> = [], action: Action) => {
  switch (action.type){
    case 'EMPLOYEE:CREATE__ERROR':
      return action.payload.error;
    case 'EMPLOYEE:RESET_NEW_EMPLOYEE_FORM':
      return [];
    default:
      return state;
  }
};

export type LocalState = {
  by_id: ById,
  all_ids: Array<string>,
  is_fetching: boolean,
  updating_id: number,
  is_creating: boolean,
  create_employee_errors: Array<string>,
  successfully_created: boolean
};

const employeeReducer = combineReducers({
  by_id: byIdReducer,
  all_ids: allIdsReducer,
  is_fetching: isFetchingReducer,
  updating_id: updatingIdReducer,
  is_creating: isCreatingReducer,
  create_employee_errors: createEmployeeErrorsReducer,
  successfully_created: successfullyCreatedReducer
});

export default employeeReducer;
