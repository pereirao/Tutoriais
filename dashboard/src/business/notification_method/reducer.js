// @flow

import _ from 'lodash';
import { combineReducers } from 'redux';
import I18n from '../../localization';
import type { Order } from './index';
import type { Action } from '../store';

type ById = {[id: number]: Order};

export const byIdReducer = (state: ById = {}, action: Action) => {
  switch (action.type){
    case 'NOTIFICATION_METHOD:FETCH__SUCCESS':
      return action.payload.entities.notification_method || {};
    case 'NOTIFICATION_METHOD:CREATE__SUCCESS':
    case 'NOTIFICATION_METHOD:UPDATE__SUCCESS':
      return {...state, ...action.payload.entities.notification_method};
    case 'NOTIFICATION_METHOD:DESTROY__SUCCESS':
      return _.omit(state, action.payload.result);
    default:
      return state;
  }
};

export const allIdsReducer = (state: Array<string> = [], action: Action) => {
  switch (action.type){
    case 'NOTIFICATION_METHOD:FETCH__SUCCESS':
      return _.uniq([...state, ...action.payload.result]);
    case 'NOTIFICATION_METHOD:CREATE__SUCCESS':
      return [...state, action.payload.result];
    case 'NOTIFICATION_METHOD:DESTROY__SUCCESS':
      return _.without(state, action.payload.result);
    default:
      return state;
  }
};

export const isFetchingReducer = (state: boolean = false, action: Action) => {
  switch (action.type){
    case 'NOTIFICATION_METHOD:FETCH__LOADING':
      return true;
    case 'NOTIFICATION_METHOD:FETCH__SUCCESS':
    case 'NOTIFICATION_METHOD:FETCH__ERROR':
      return false;
    default:
      return state;
  }
};

export const updatingIdReducer = (state: number = null, action: Action) => {
  switch (action.type){
    case 'NOTIFICATION_METHOD:UPDATE__LOADING':
      return action.meta.id;
    case 'NOTIFICATION_METHOD:UPDATE__SUCCESS':
    case 'NOTIFICATION_METHOD:UPDATE__ERROR':
      return null;
    default:
      return state;
  }
};

export const isCreatingReducer = (state: boolean = false, action: Action) => {
  switch (action.type){
    case 'NOTIFICATION_METHOD:CREATE__LOADING':
      return true;
    case 'NOTIFICATION_METHOD:CREATE__SUCCESS':
    case 'NOTIFICATION_METHOD:CREATE__ERROR':
      return false;
    default:
      return state;
  }
};

export const successfullyCreatedReducer = (state: boolean = false, action: Action) => {
  switch (action.type){
    case 'NOTIFICATION_METHOD:CREATE__SUCCESS':
      return true;
    case 'NOTIFICATION_METHOD:RESET_NEW_NOTIFICATION_METHOD_FORM':
    case 'NOTIFICATION_METHOD:CREATE__ERROR':
      return false;
    default:
      return state;
  }
};

export const createNotificationMethodErrorsReducer = (state: Array<string> = [], action: Action) => {
  switch (action.type){
    case 'NOTIFICATION_METHOD:CREATE__ERROR':{
      if (action.payload.error === 'Internal Server Error'){
        return [I18n.t('form.error.invalid')];
      } else {
        return state;
      }
    }
    case 'NOTIFICATION_METHOD:RESET_NEW_NOTIFICATION_METHOD_FORM':
      return [];
    default:
      return state;
  }
};

export type LocalState = {
  by_id: ById,
  all_ids: Array<string>,
  is_fetching: boolean,
  updating_id: number,
  is_creating: boolean,
  create_notification_method_errors: Array<string>,
  successfully_created: boolean
};

const notificationMethodReducer = combineReducers({
  by_id: byIdReducer,
  all_ids: allIdsReducer,
  is_fetching: isFetchingReducer,
  updating_id: updatingIdReducer,
  is_creating: isCreatingReducer,
  create_notification_method_errors: createNotificationMethodErrorsReducer,
  successfully_created: successfullyCreatedReducer
});

export default notificationMethodReducer;
