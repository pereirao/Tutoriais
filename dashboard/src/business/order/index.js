// @flow

export type OrderTags = {
  gift: boolean,
  out_of_hours: boolean,
  vip: boolean,
  corporate: boolean,
  scheduled: boolean,
  new_customer: boolean
};

export type BillingInfo = {
  name: string,
  card_type: string,
  card_last4: string
};

export type Address = {
  address1: string,
  address2: string,
  city: string,
  state: string,
  zip_code: string,
  coords: {lat: number, lng: number}
};

export type PickupDetail = {
  id: string,
  name: string,
  phone: string
};

export type Amounts = {
  total: number,
  subtotal: number,
  tax: number,
  tip: number,
  minibar_promos: number,
  store_discounts: number,
  discounts?: number, // FIXME: optional until sunset
  delivery_fee: number,
};

export type DeliveryService = {
  id: number,
  name: string,
  email: string,
  created_at: string,
  updated_at: string,
  minimal_delivery_time: number
}

export type DeliveryMethodType = 'on_demand' | 'shipped' | 'pickup';

export type DeliveryMethod = {
  type: DeliveryMethodType,
  expectation: string
};

export type AttributeFilters = 'vip' | 'gift' | 'scheduled' | 'exception';

export type OrderFilters = {
  date_range: {
    start: string, // iso-8601 timestamp
    end: string // iso-8601 timestamp
  },
  delivery_method_types: Array<DeliveryMethodType>,
  attributes: Array<AttributeFilters>
}

export type ExceptionMetadata = {
  order_item_sku?: string,
  replacement_item?: string,
  replacement_price?: string,
  correct_price?: string,
  incorrect_listing?: string,
  incomplete_address?: string,
  suspected_fraud?: string,
  failed_delivery?: string,
  driver_unavailable_why?: string,
  other?: string,
};

export type ExceptionType = 'out_of_stock' |
  'incorrect_pricing' |
  'incorrect_listing' |
  'incomplete_address' |
  'suspected_fraud' |
  'driver_unavailable' |
  'failed_delivery' |
  'other';

export type Exception = {
  type: ExceptionType,
  description: string,
  metadata: ExceptionMetadata
};

export const PROBLEM_TYPES = [
  'out_of_stock',
  'incorrect_pricing',
  'incorrect_listing',
  'incomplete_address',
  'suspected_fraud',
  'driver_unavailable',
  'failed_delivery',
  'other'
];

export type OrderListName = 'all' | 'active' | 'completed' | 'future';

export type RecipientInfo = { short_name: string, long_name: string, phone: string };

export type TimelineEntryType = 'created_at'
  | 'confirmed_at'
  | 'en_route_at'
  | 'delivered_at'
  | 'canceled_at'
  | 'comment';

export type TimelineEntry = {
  time: string,
  type: TimelineEntryType,
  meta?: { name?: string, driverName?: string }
}

export type OrderState = 'paid'
  | 'confirmed'
  | 'scheduled'
  | 'canceled'
  | 'exception'
  | 'en_route'
  | 'delivered';

export type OrderItem = {
  id: number,
  sku: string,
  name: string,
  item_size: string,
  unit_price: number,
  quantity: number
};

export type Order = {
  id: number,
  number: number,
  created_at: string, // rails timestamp
  confirmed_at: string, // rails timestamp
  canceled_at: string, // rails timestamp
  delivered_at: string, // rails timestamp
  scheduled_for: string, // rails timestamp
  scheduled_for_end: string, // rails timestamp
  order_time: string, // rails timestamp
  birthdate: string,
  notes: string,
  delivery_method: DeliveryMethod,
  state: OrderState,
  gift_message: string,
  recipient_info: RecipientInfo,
  customer_name: string,
  billing: BillingInfo,
  address?: Address,
  pickup_detail?: PickupDetail,
  amounts: Amounts,
  type_tags: OrderTags,
  receipt_url: string,
  order_items: Array<OrderItem>,
  comment_ids: Array<number>, // NOTE: shouldn't be hydrated
  order_adjustment_ids: Array<number>, // NOTE: shouldn't be hydrated
  driver_id: number,
  driver?: any,
  using_delivery_service: boolean,
  delivery_service: DeliveryService
};

export * as order_actions from './actions';
export * as order_helpers from './helpers';
export * as order_epics from './epics';
export { default as orderReducer } from './reducer';
export { default as order_selectors } from './selectors';
