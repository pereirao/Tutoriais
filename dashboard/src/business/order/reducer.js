// @flow

import _ from 'lodash';
import { combineReducers } from 'redux';
import type { Order } from './index';
import type { Action } from '../store';

type ById = {[id: number]: Order};

export const byIdReducer = (state: ById = {}, action: Action) => {
  switch (action.type){
    case 'ORDER:FETCH_ORDER__SUCCESS':
    case 'ORDER:UPDATE_ORDER__SUCCESS':
    case 'ORDER:FETCH_ACTIVE__SUCCESS':
    case 'ORDER:FETCH_COMPLETED__SUCCESS':
    case 'ORDER:FETCH_FUTURE__SUCCESS':
    case 'ORDER:FETCH_ALL__SUCCESS':
      return {...state, ...(action.payload.entities.order || {})};
    case 'ADJUSTMENT:FETCH__SUCCESS':{
      const { result } = action.payload;
      const { order_id } = action.meta;
      const updated_order_by_id = {[order_id]: {...state[order_id], order_adjustment_ids: result}};
      return { ...state, ...updated_order_by_id };
    }
    case 'COMMENT:ADD__SUCCESS':{
      const { result } = action.payload;
      const { order_id } = action.meta;
      const updated_order_by_id = {[order_id]: {...state[order_id], comment_ids: [...state[order_id].comment_ids, result] }};
      return { ...state, ...updated_order_by_id };
    }
    case 'COMMENT:FETCH__SUCCESS':{
      const { result } = action.payload;
      const { order_id } = action.meta;
      const updated_order_by_id = {[order_id]: {...state[order_id], comment_ids: result}};
      return { ...state, ...updated_order_by_id };
    }
    default:
      return state;
  }
};

export const isFetchingReducer = (state: boolean = false, action: Action) => {
  switch (action.type){
    case 'ORDER:FETCH_ACTIVE__LOADING':
    case 'ORDER:FETCH_COMPLETED__LOADING':
    case 'ORDER:FETCH_FUTURE__LOADING':
    case 'ORDER:FETCH_ALL__LOADING':
      return true;
    case 'ORDER:FETCH_ACTIVE__SUCCESS':
    case 'ORDER:FETCH_ACTIVE__ERROR':
    case 'ORDER:FETCH_COMPLETED__SUCCESS':
    case 'ORDER:FETCH_COMPLETED__ERROR':
    case 'ORDER:FETCH_FUTURE__SUCCESS':
    case 'ORDER:FETCH_FUTURE__ERROR':
    case 'ORDER:FETCH_ALL__SUCCESS':
    case 'ORDER:FETCH_ALL__ERROR':
      return false;
    default:
      return state;
  }
};

export const isUpdatingReducer = (state: boolean = false, action: Action) => {
  switch (action.type){
    case 'ORDER:UPDATE_ORDER__LOADING':
      return true;
    case 'ORDER:UPDATE_ORDER__SUCCESS':
    case 'ORDER:UPDATE_ORDER__ERROR':
      return false;
    default:
      return state;
  }
};

export const EMPTY_FILTERS = {
  date_range: {start: null, end: null},
  delivery_method_types: [],
  attributes: []
};

type OrderList = {
  ids: Array<string>,
  total_count: number,
  total_pages: number,
  next_page: number,
  query: string,
  filters: OrderFilters,
  is_stale: boolean
};
export const makeOrderListReducers = (update_action_type: string) => {
  const idsReducer = (state: Array<string> = [], action: Action) => {
    switch (action.type){
      case update_action_type:{
        if (action.meta.page === 1) return action.payload.result.orders;
        return _.uniq([...state, ...action.payload.result.orders]);
      }
      default:
        return state;
    }
  };
  const totalCountReducer = (state: number = 0, action: Action) => {
    switch (action.type){
      case update_action_type:
        return action.payload.result.total_count;
      default:
        return state;
    }
  };
  const totalPagesReducer = (state: number = 1, action: Action) => {
    switch (action.type){
      case update_action_type:
        return action.payload.result.total_pages;
      default:
        return state;
    }
  };
  const nextPageReducer = (state: number = 1, action: Action) => {
    switch (action.type){
      case update_action_type:
        return action.payload.result.total_count === 0 ? action.meta.page : action.meta.page + 1;
      default:
        return state;
    }
  };
  const queryReducer = (state: string = '', action: Action) => {
    switch (action.type){
      case update_action_type:
        return _.get(action.meta, 'query') || '';
      default:
        return state;
    }
  };
  const filtersReducer = (state: OrderFilters = EMPTY_FILTERS, action: Action) => {
    switch (action.type){
      case update_action_type:
        return _.get(action.meta, 'filters') || state;
      default:
        return state;
    }
  };
  const isStaleReducer = (state: boolean = false, action: Action) => {
    switch (action.type){
      case 'ORDER:MARK_LISTS_STALE':
        return true;
      case update_action_type:{
        if (action.meta.page !== 1) return state; //still stale if not fetching first page
        return false;
      }
      default:
        return state;
    }
  };

  return {
    ids: idsReducer,
    total_pages: totalPagesReducer,
    total_count: totalCountReducer,
    next_page: nextPageReducer,
    query: queryReducer,
    filters: filtersReducer,
    is_stale: isStaleReducer
  };
};

export const activeReducers = makeOrderListReducers('ORDER:FETCH_ACTIVE__SUCCESS');
export const allReducers = makeOrderListReducers('ORDER:FETCH_ALL__SUCCESS');
export const completedReducers = makeOrderListReducers('ORDER:FETCH_COMPLETED__SUCCESS');
export const futureReducers = makeOrderListReducers('ORDER:FETCH_FUTURE__SUCCESS');

export type LocalState = {
  by_id: ById,
  all: OrderList,
  active: OrderList,
  completed: OrderList,
  future: OrderList,
  is_fetching: boolean,
  is_updating: boolean
};

const orderReducer = combineReducers({
  by_id: byIdReducer,
  all: combineReducers(allReducers),
  active: combineReducers(activeReducers),
  completed: combineReducers(completedReducers),
  future: combineReducers(futureReducers),
  is_fetching: isFetchingReducer,
  is_updating: isUpdatingReducer
});

export default orderReducer;
