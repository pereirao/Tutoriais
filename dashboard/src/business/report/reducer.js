// @flow

import { combineReducers } from 'redux';
import type { Report } from './index';
import type { Action } from '../store';

type ById = {[id: number]: Report};

export const byIdReducer = (state: ById = {}, action: Action) => {
  switch (action.type){
    case 'REPORT:FETCH__SUCCESS':
    case 'REPORT:POLL__SUCCESS':
      return {...state, ...(action.payload.entities.report || {})};
    case 'REPORT:CREATE__SUCCESS':
      return { ...action.payload.entities.report, ...state };
    default:
      return state;
  }
};

export const allIdsReducer = (state: Array<string> = [], action: Action) => {
  switch (action.type){
    case 'REPORT:FETCH__SUCCESS':
    case 'REPORT:POLL__SUCCESS':
      return action.payload.result;
    case 'REPORT:CREATE__SUCCESS':
      return [action.payload.result, ...state];
    default:
      return state;
  }
};

export type LocalState = {
  by_id: ById,
  all_ids: Array<string>
};

const reportReducer = combineReducers({
  by_id: byIdReducer,
  all_ids: allIdsReducer
});

export default reportReducer;
