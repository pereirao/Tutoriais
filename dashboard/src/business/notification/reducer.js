//@flow

import { combineReducers } from 'redux';
import _ from 'lodash';
import { order_helpers } from '../order';
import { comment_helpers } from '../comment';
import type { Notification } from './index';
import type { Action } from '../store';

const { isUnconfirmed, isCanceled } = order_helpers;
const { commentNotification } = comment_helpers;

type ById = {[id: string]: Notification};

export const byIdReducer = (state: ById = {}, action: Action) => {
  switch (action.type){
    case 'NOTIFICATION:PUSH':{
      const { id, order_id, notification_type } = action.payload;
      return { ...state, [id]: {id, order_id, notification_type} };
    }
    case 'SESSION:PING__SUCCESS':{
      const unconfirmed_orders = action.payload.unconfirmed_shipments.map(shipment_id => {
        const id = `unconfirmed_${shipment_id}`;
        return { id, order_id: shipment_id, notification_type: 'unconfirmed' };
      });
      return {...state, ..._.keyBy(unconfirmed_orders, 'id')};
    }
    case 'ORDER:FETCH_ORDER__SUCCESS':
    case 'ORDER:UPDATE_ORDER__SUCCESS':{
      const order = _.values(action.payload.entities.order)[0];
      if (isUnconfirmed(order)){
        const id = `unconfirmed_${order.id}`;
        return { ...state, [id]: { id, order_id: order.id, notification_type: 'unconfirmed' } };
      } else if (isCanceled(order) && _.get(action.meta, 'notify')){
        const id = `canceled_${order.id}`;
        return { ...state, [id]: { id, order_id: order.id, notification_type: 'canceled' } };
      }
      return _.omit(state, order.id);
    }
    case 'COMMENT:FETCH__SUCCESS':{
      if (action.meta.from_order_detail) return state; // don't notify for comments fetched while on that order detail screen
      if (_.get(action.payload.entities.comment, 'author.email') === action.meta.my_email){
        return state; // don't notify for comments added by "me"
      }
      const comment_notification = commentNotification(action.payload.entities.comment, action.meta.order_id);
      return { ...state, [comment_notification.id]: comment_notification};
    }
    case 'NOTIFICATION:REMOVE':
      return _.omit(state, action.payload.notification_id);
    default:
      return state;
  }
};

export const displayOrderReducer = (state: Array<string> = [], action: Action) => {
  switch (action.type){
    case 'NOTIFICATION:PUSH':
      return [
        ...state,
        action.payload.id
      ];
    case 'SESSION:PING__SUCCESS':{
      const unconfirmed_ids = action.payload.unconfirmed_shipments.map(shipment_id => `unconfirmed_${shipment_id}`);
      const cleaned_prev_state = state.filter(notification_id => (
        unconfirmed_ids.includes(notification_id) || !notification_id.startsWith('unconfirmed') // remove stale unconfirmed notifications 
      ));
      return _.uniq([...cleaned_prev_state, ...unconfirmed_ids]);
    }
    case 'ORDER:FETCH_ORDER__SUCCESS':
    case 'ORDER:UPDATE_ORDER__SUCCESS':{
      const order = _.values(action.payload.entities.order)[0];
      if (isUnconfirmed(order)){
        const notification_id = `unconfirmed_${action.payload.result}`;
        return _.uniq([...state, notification_id]);// if unconfirmed append to end if new, otherwise do nothing
      } else if (isCanceled(order) && _.get(action.meta, 'notify')){
        const notification_id = `canceled_${action.payload.result}`;
        return _.uniq([..._.without(state, `unconfirmed_${action.payload.result}`), notification_id]);
      } else {
        return _.without(state, `unconfirmed_${action.payload.result}`); // if there isn't a notification after update remove if present
      }
    }
    case 'COMMENT:FETCH__SUCCESS':{
      if (action.meta.from_order_detail) return state; // don't notify for comments fetched while on that order detail screen
      return [...state, commentNotification(action.payload.entities.comment, action.meta.order_id).id];
    }
    case 'NOTIFICATION:REMOVE':
      return _.without(state, action.payload.notification_id);
    default:
      return state;
  }
};

export type LocalState = {
  by_id: ById,
  display_order: Array<string>
};

const notificationReducer = combineReducers({
  by_id: byIdReducer,
  display_order: displayOrderReducer
});

export default notificationReducer;
