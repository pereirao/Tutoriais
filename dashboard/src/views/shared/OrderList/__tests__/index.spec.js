import React from 'react';

// NOTE: the named export isn't a connected component
import { OrderList } from '../index';
import { EMPTY_FILTERS } from '../../../../business/order/reducer';
import order_factory from '../../../../business/order/__tests__/order.factory';

it('renders correctly', () => {
  expect(
    shallow(<OrderList
      fetch={() => {}}
      applied_filters={EMPTY_FILTERS}
      orders={[order_factory.build({id: 1}), order_factory.build({id: 2})]}
      list_name="active" />
    )
  ).toMatchSnapshot();
});
