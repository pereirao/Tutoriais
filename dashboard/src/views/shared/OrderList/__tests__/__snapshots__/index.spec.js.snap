// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <div
    className="actionRow"
  >
    <SearchInput
      onChange={[Function]}
      placeholder="Search name, order number, address"
      submit={[Function]}
    />
    <OrderFiltersModal
      applyFilters={[Function]}
      are_applied_filters={false}
      filters={
        Object {
          "attributes": Array [],
          "date_range": Object {
            "end": null,
            "start": null,
          },
          "delivery_method_types": Array [],
        }
      }
      setDateFilter={[Function]}
      toggleFilter={[Function]}
      toggleVisibility={[Function]}
      visible={false}
    />
  </div>
  <RefreshList
    refresh={[Function]}
  />
  <AppliedFilters
    clearFilters={[Function]}
    filters={Array []}
    removeFilter={[Function]}
  />
  <Connect(OrderTable)
    orders={
      Array [
        Object {
          "address": Object {
            "address1": "21 Jump Street",
            "address2": "Apt 5",
            "city": "Brooklyn",
            "coords": Object {
              "lat": 40.719507,
              "lng": -73.941461,
            },
            "state": "NY",
            "zip_code": 11206,
          },
          "amounts": Object {
            "delivery_fee": 5,
            "minibar_promos": 2.35,
            "store_discounts": 2,
            "subtotal": 43.47,
            "tax": 3.48,
            "tip": 4.35,
            "total": 51.95,
          },
          "billing": Object {
            "cc_type": "Amex",
            "last_digits": "4321",
          },
          "comment_ids": Array [],
          "customer_name": "Biggie Smalls",
          "delivery_method": Object {
            "maximum_delivery_expectation": 60,
            "type": "on_demand",
          },
          "driver_id": 15,
          "id": 1,
          "notes": "Ring the doorbell",
          "number": 1,
          "order_items": Array [],
          "order_time": "2017-01-01T00:00:00-05:00",
          "receipt_url": "http://www.minibardelivery.com",
          "recipient_info": Object {
            "long_name": "Biggie Smalls - East Coast",
            "phone": "5555555555",
            "short_name": "Biggie Smalls",
          },
          "state": "paid",
          "type_tags": Object {
            "corporate": false,
            "gift": false,
            "out_of_hours": false,
            "scheduled": false,
            "vip": true,
          },
        },
        Object {
          "address": Object {
            "address1": "21 Jump Street",
            "address2": "Apt 5",
            "city": "Brooklyn",
            "coords": Object {
              "lat": 40.719507,
              "lng": -73.941461,
            },
            "state": "NY",
            "zip_code": 11206,
          },
          "amounts": Object {
            "delivery_fee": 5,
            "minibar_promos": 2.35,
            "store_discounts": 2,
            "subtotal": 43.47,
            "tax": 3.48,
            "tip": 4.35,
            "total": 51.95,
          },
          "billing": Object {
            "cc_type": "Amex",
            "last_digits": "4321",
          },
          "comment_ids": Array [],
          "customer_name": "Biggie Smalls",
          "delivery_method": Object {
            "maximum_delivery_expectation": 60,
            "type": "on_demand",
          },
          "driver_id": 15,
          "id": 2,
          "notes": "Ring the doorbell",
          "number": 2,
          "order_items": Array [],
          "order_time": "2017-01-01T00:00:00-05:00",
          "receipt_url": "http://www.minibardelivery.com",
          "recipient_info": Object {
            "long_name": "Biggie Smalls - East Coast",
            "phone": "5555555555",
            "short_name": "Biggie Smalls",
          },
          "state": "paid",
          "type_tags": Object {
            "corporate": false,
            "gift": false,
            "out_of_hours": false,
            "scheduled": false,
            "vip": true,
          },
        },
      ]
    }
    updateable={true}
  />
  <Footer
    empty_message="No Active Orders"
    has_more={false}
    is_empty={false}
    loadMore={[Function]}
  />
</div>
`;
